### ISSUE TITLE
Error with CanonicalAlgorithm implementation

### ISSUE THREAD

**Issue by microshine** [yaronn/xml-crypto#102]
``` javascript
CanonicalizationAlgorithms  = {
        "http://www.w3.org/2001/10/xml-exc-c14n#": ExclusiveCanonicalization,
        "http://www.w3.org/2001/10/xml-exc-c14n#WithComments": ExclusiveCanonicalizationWithComments,
        "http://www.w3.org/2000/09/xmldsig#enveloped-signature": EnvelopedSignature
    };
```

All algorithms have got `process` function, but with different result types

``` javascript
EnvelopedSignature.process(node: Node, options?): Node;
ExclusiveCanonicalization.process(node: Node, options?): string;
ExclusiveCanonicalizationWithComments.process(node: Node, options?): string;
```

It can be a problem with `createSignature` function
https://github.com/yaronn/xml-crypto/blob/master/lib/signed-xml.js#L701-L704
because `signatureAlgorithm.getSignature` needs `signedInfo: string | Buffer`

**Comment by yaronn** [comment]
You are right that this is confusing and bad coding. I don't expect this to be a problem in practice since an EnvelopedSignature does not make sense as a canonicalization algorithm for the signature. But we should definitely get rid of this when we have a chance.

**Comment by microshine** [comment]
The same error with 

```javascript
var canonizedSignedInfo = canAlgorithm.process(node);
var signatureAlgorithm = this.findSignatureAlgorithm(this.signatureAlgorithm);
this.signatureValue = signatureAlgorithm.getSignature(canonizedSignedInfo, this.signingKey);
```

`process` - can return Node
`getSignature` - 1 param must be string

**Comment by rmhrisk** [comment]
We ended up fixing this in our implementation derived from this great package :  https://www.npmjs.com/package/xadesjs

**Comment by avckp** [comment]
Any pointers on how to fix this?
I do not have understanding of what the node is, but I would like to make the return type `string` for process method of the `enveloped` algorithm as I have immediate need for the same.
Otherwise are their any known mature implementations which do the same?

**Comment by yaronn** [comment]
@avckp can you share the code that consumes xml-crypto? I want to understand the sequence of algorithms.

**Comment by avckp** [comment]
@yaronn I need to generate the signature in the following format
```
<Signature xmlns="http://www.w3.org/2000/09/xmldsig#">
	<SignedInfo>
		<CanonicalizationMethod Algorithm="http://www.w3.org/TR/2001/REC-xml-c14n-20010315" />
		<SignatureMethod Algorithm="http://www.w3.org/2000/09/xmldsig#rsa-sha1" />
		<Reference URI="">
			<Transforms>
				<Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature" />
			</Transforms>
			<DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha1" />
			<DigestValue>2QvzuMcZU/D6Nbk3lmUaxupnzG8=</DigestValue>
		</Reference>
	</SignedInfo>
	<SignatureValue>R5TfpMgiKT7...</SignatureValue>
	<KeyInfo>
		<X509Data>
			<X509SubjectName>....</X509SubjectName>
	<X509Certificate>.....</X509Certificate>
		</X509Data>
	</KeyInfo>
</Signature>
```
My code looks like
```
 var sig = new SignedXml();
      sig.addReference("//*[local-name(.)='Hmac']");
      sig.signingKey = fs.readFileSync("certs/keyfile.key.pem");
      sig.keyInfoProvider = new MyKeyInfo();
      sig.canonicalizationAlgorithm = "http://www.w3.org/2000/09/xmldsig#enveloped-signature";
      sig.computeSignature(xml);
```
Signing with this algorithm was failing (as enveloped-signature returns node in place of string), I modified the `SignedXml.prototype.createSignature` to use 

`706  this.signatureValue = signatureAlgorithm.getSignature(**canonizedSignedInfo.toString()**, this.signingKey)`

While it gives me output, I am not sure what I am doing here.
Am I supposed to convert the node to string inside the `transform.process` itself?

Also my output has 
`<CanonicalizationMethod Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>`
and 
`<Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>`
I am expecting it to be
`<Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature" />`

**Comment by yaronn** [comment]
I think you mixed the algorithms you need. The right place to add the enveloped transformation is here:

      sig.addReference("//*[local-name(.)='Hmac']", ["http://www.w3.org/2000/09/xmldsig#enveloped-signature"]);

As for the canonicalization algorithm (http://www.w3.org/TR/2001/REC-xml-c14n-20010315), that algorithm is not supported yet. The documentation explains how to add algorithms if you want but it is sometimes time consuming. It is not always true, so not sure I would recommend it for produciton use, but I think in many cases that algorithm is the same as exclusive canonicalization so for testing you could try to do this:

    SignedXml.CanonicalizationAlgorithms['http://www.w3.org/TR/2001/REC-xml-c14n-20010315'] = SignedXml.ExclusiveCanonicalization;
    sig.canonicalizationAlgorithm = "http://www.w3.org/TR/2001/REC-xml-c14n-20010315";

**Comment by FraGoTe** [comment]
@yaronn  I added as you recommended but I got the following error:

Possibly unhandled Error: canonicalization algorithm 'http://www.w3.org/TR/2001/REC-xml-c14n-20010315' is not supported
    at SignedXml.findCanonicalizationAlgorithm (/Users/FraGoTe/Documents/Projects/xxx/node_modules/xml-crypto/lib/signed-xml.js:292:14)

**Comment by FraGoTe** [comment]
I did :
```
 SignedXml.CanonicalizationAlgorithms['http://www.w3.org/TR/2001/REC-xml-c14n-20010315'] = SignedXml.CanonicalizationAlgorithms['http://www.w3.org/2001/10/xml-exc-c14n#'];

    sig.canonicalizationAlgorithm = 'http://www.w3.org/TR/2001/REC-xml-c14n-20010315';
```

But i got the same <ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>

**Comment by yaronn** [comment]
@FraGoTe make sure the first line you pasted is called before the second one. You can also try to hard code that algorithm here just to test https://github.com/yaronn/xml-crypto/blob/master/lib/signed-xml.js#L225

**Comment by vini-guerrero** [comment]
Could anyone run it for that algorithm?

 **sig.canonicalizationAlgorithm = 'http://www.w3.org/TR/2001/REC-xml-c14n-20010315';**

Any tips how to apply this?

**Comment by LoneRifle** [comment]
Fixed in #116
